python/lsst/pipe/base/script/transfer_from_graph.py

```diff
@@ -0,0 +1,138 @@
+    register_dataset_types : `bool`
+        Indicate whether missing dataset types should be registered.
+    transfer_dimensions : `bool`
+        Indicate whether dimensions should be transferred along with datasets.
+        It can be more efficient to disable this if it is known that all
+        dimensions exist.
+    update_output_chain : `bool`
+        If quantum graph metadata includes output run name and output
+        collection which is a chain, update the chain definition to include run
+        name as a the first collection in the chain.
+
+    Returns
+    -------
+    count : `int`
+        Actual count of transferred datasets.
+    """
+
+    # Read whole graph into memory
+    qgraph = QuantumGraph.loadUri(graph)
+
+    # Collect output refs that could be created by this graph.
+    output_refs: set[DatasetRef] = set(qgraph.globalInitOutputRefs())
+    for task_def in qgraph.iterTaskGraph():
+        if refs := qgraph.initOutputRefs(task_def):
+            output_refs.update(refs)
+    for qnode in qgraph:
+        for refs in qnode.quantum.outputs.values():
+            output_refs.update(refs)
+
+    # Make QBB, its config is the same as output Butler.
+    qbb = QuantumBackedButler.from_predicted(
+        config=dest,
+        predicted_inputs=[ref.getCheckedId() for ref in output_refs],
+        predicted_outputs=[],
+        dimensions=qgraph.universe,
+        datastore_records={},
+    )
+
+    dest_butler = Butler(dest, writeable=True)
+
+    transferred = dest_butler.transfer_from(
+        qbb,
+        output_refs,
+        transfer="auto",
+        register_dataset_types=register_dataset_types,
+        transfer_dimensions=transfer_dimensions,
+    )
+    count = len(transferred)
+
+    # If anything was transferred then update output chain definition if asked.
+    if count > 0 and update_output_chain and (metadata := qgraph.metadata) is not None:
+        # These are defined in CmdLineFwk.
+        output_run = metadata.get("output_run")
+        output = metadata.get("output")
+        input = metadata.get("input")
+        if output_run is not None and output is not None:
+            _update_chain(dest_butler.registry, output, output_run, input)
+
+    return count
+
+
+def _update_chain(registry: Registry, output_chain: str, output_run: str, inputs: list[str] | None) -> None:
+    """Update chain definition if it exists to include run as the first item
+    in a chain. If it does not exist then create it to include all inputs and
+    output.
+    """
+    try:
+        chain_definition = list(registry.getCollectionChain(output_chain))
+    except CollectionTypeError:
+        # Not a chained collection, ignore it.
+        # TODO: Print a warning message?
+        return
+    except MissingCollectionError:
+        # This probably means that someone forgot to make output chain
+        # collection, we want to create it to include inputs and output run
+        # (this reproduces logic in CmdLineFwk).
```

Comment:
I don't think this is a case of "forgot" - it's just the common case where this is the first processing for that particular chain, as we're careful to avoid creating the chain until we have something to actually put in it.